components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/wireguard_tunnel_service.cc

```diff
@@ -0,0 +1,227 @@
+  }
+  ScopedScHandle service(
+      ::OpenService(scm.Get(), GetBraveVpnWireguardTunnelServiceName().c_str(),
+                    SERVICE_ALL_ACCESS));
+
+  if (service.IsValid()) {
+    if (IsServiceRunning(service.Get())) {
+      SERVICE_STATUS stt;
+      if (!ControlService(service.Get(), SERVICE_CONTROL_STOP, &stt)) {
+        VLOG(1) << "ControlService failed to send stop signal";
+        return false;
+      }
+    }
+    if (!DeleteService(service.Get())) {
+      VLOG(1) << "DeleteService failed";
+      return false;
+    }
+  }
+  return true;
+}
+
+// Creates and launches a new Wireguard service with specific config.
+bool CreateAndRunBraveWireguardService(const std::wstring& config) {
+  ScopedScHandle scm(::OpenSCManager(nullptr, nullptr, SC_MANAGER_ALL_ACCESS));
+  if (!scm.IsValid()) {
+    VLOG(1) << "::OpenSCManager failed. service_name: "
+            << GetBraveVpnWireguardTunnelServiceName()
+            << ", error: " << std::hex << HRESULTFromLastError();
+    return false;
+  }
+  base::FilePath directory;
+  if (!base::PathService::Get(base::DIR_EXE, &directory)) {
+    return false;
+  }
+  base::CommandLine service_cmd(
+      directory.Append(brave_vpn::kBraveVpnWireguardServiceExecutable));
+  service_cmd.AppendSwitchNative(
+      brave_vpn::kBraveVpnWireguardServiceConnectSwitchName, config);
+
+  ScopedScHandle service(::CreateService(
+      scm.Get(), GetBraveVpnWireguardTunnelServiceName().c_str(),
+      GetBraveVpnWireguardTunnelServiceName().c_str(), SERVICE_ALL_ACCESS,
+      SERVICE_WIN32_OWN_PROCESS, SERVICE_DEMAND_START, SERVICE_ERROR_NORMAL,
+      service_cmd.GetCommandLineString().c_str(), NULL, NULL, L"Nsi\0TcpIp\0",
+      NULL, NULL));
+  if (!service.IsValid()) {
+    VLOG(1) << "::CreateService failed. service_name: "
+            << GetBraveVpnWireguardTunnelServiceName() << ", error: 0x"
+            << ::GetLastError();
+    return false;
+  }
+
+  SERVICE_SID_INFO info;
+  info.dwServiceSidType = SERVICE_SID_TYPE_UNRESTRICTED;
+  if (!ChangeServiceConfig2(service.Get(), SERVICE_CONFIG_SERVICE_SID_INFO,
+                            &info)) {
+    VLOG(1) << "Failed to configure service 0x" << std::hex
+            << HRESULTFromLastError();
+    return false;
+  }
+
+  if (!StartService(service.Get(), 0, NULL)) {
+    VLOG(1) << "Failed to start service 0x" << std::hex
+            << HRESULTFromLastError();
+    return false;
+  }
+
+  return DeleteService(service.Get()) != 0;
+}
+
+int RunWireguardTunnelService(const std::wstring& encoded_config) {
+  std::string config;
+  if (!base::Base64Decode(base::WideToUTF8(encoded_config), &config) ||
+      config.empty()) {
+    VLOG(1) << "Unable to decode wireguard config";
+    return S_FALSE;
+  }
+  base::ScopedTempDir temp_dir;
+  if (!temp_dir.CreateUniqueTempDir()) {
+    return S_FALSE;
+  }
+
+  base::FilePath temp_file_path(
+      temp_dir.GetPath().Append(kBraveWireguardConfig));
+  if (!base::WriteFile(temp_file_path, config)) {
```

Comment:
What file permissions do we set on this file? Is it readable by all users or just the current user?

Reply:
readable for all now

Reply:
I forget how Windows permissions work, but is there an equivalent to UNIX's `-rw-------` (or `-rw-rw----`)?

If so, it would be good to avoid making this a world-readable file since it contains credentials.

Reply:
done in https://github.com/brave/brave-core/pull/18513/commits/4b5fee0d45e9d37990d29969ce9f0d8f6f7d3b97, made it possible to read/write/delete only by Windows Services.